README
$Id: README,v 1.6 2006-06-30 06:25:08 oops Exp $

libkrisp - Korea ISP library

작성: 김정균 2006.06.09 http://www.oops.org

libkrisp 는 한국의 ip address 에 대한 ISP 정보를 얻기 위한 라이브러리이다.
이 library 를 사용하기 위해서는 sqlite library 가 필요하며, sqlite2 또는 sqlite3
이 필요하다. sqlite2 와 sqlite3 이 동시에 설치가 되어 있으면, sqlite3 을 이용
하며, sqlite3 을 사용하는 것을 권장한다. (sqlite2 는 가끔 오동작을 한다.)

간단한 사용법은 아래와 같다. (자세한 사항은 'man 3 kr_search' 를 참고한다.)

#include <krisp.h>
extern char dberr[1024];

int main (void) {
    KR_API db;
    KRNET_API isp;
    char *ip = "168.126.63.1";

	db.gi = NULL;

    if ( kr_open (&db) ) {
        fprintf (stderr, "ERROR Connect: %s\n", dberr);
        return 1;
    }

	strcpy (isp.ip, ip);
    kr_search (&isp, &db, NULL);

	kr_close (&db);

    printf ("%-15s %-15s %-20s %-15s %-15s %s\n",
            ip, isp.ip, isp.icode, isp.network, isp.broadcast, isp.iname);

    return 0;
}

또한, configure 시에 --with-geoip 옵션을 이용하여 geoip 를 연동할 수 있다.
geoip 를 연동하면 국가 코드와 국가 이름이 netinfos structure 에 추가 된다.
netinfos structure 는 다음과 같이 구성이 되어 있다.

struct netinfos {
    char            key[16];
    char            ip[16];
    char            netmask[16];
    char            network[16];
    char            broadcast[16];
    char            icode[128];
    char            iname[128];
#ifdef HAVE_LIBGEOIP
    char            gcode[4];
    char            gname[128];
#endif
};

geoip 를 추가하는 예제는 krisplookup.c 를 참고하기 바란다. 아주 간단한 소스
이다.


Changes
Changelog


* 3.1.6
Tue, Jan 24 2017 JoungKyun.Kim <http://oops.org>
	- applying changed format about KRNIC site reoranization on january 2017.
	- fixed PHP 7 changing about contribute scripts

Tue, Feb 17 2016 JoungKyun.Kim <http://oops.org>
	- start 3.1.6


* 3.1.5
Tue, Feb 16 2016 JoungKyun.Kim <http://oops.org>
	- fixed access uninitialized variable

Sat, Feb 15 2013 JoungKyun.Kim <http://oops.org>
	- fixed libiconv link on cygwin

Fri, Feb 14 2013 JoungKyun.Kim <http://oops.org>
	- support windows native binary
      use MinGW32 with Code::Blocks workspace
	- fixed wrong db close when database file not found

Tue, Feb 11 2013 JoungKyun.Kim <http://oops.org>
	- fixed build error and shared library on cygwin

Fri, Oct 25 2013 JoungKyun.Kim <http://oops.org>
	- start 3.1.5 tree


* 3.1.4
Fri, Oct 25 2013 JoungKyun.Kim <http://oops.org>
	- release 3.1.4
	- fixed opening the wrong path of database
	- add kr_range API for ISP or Country

Thu, Nov  1 2012 JoungKyun.Kim <http://oops.org>
	- fixed libipcalc link

Tue, Sep 18 2012 JoungKyun.Kim <http://oops.org>
	- fixed cygwin compile environments


* 3.1.3
Mon, Feb 27 2012 JoungKyun.Kim <http://oops.org>
	- release 3.1.3
	- fixed missing krversion.h

Mon, May  2 2011 JoungKyun.Kim <http://oops.org>
	- add 'extern C' for C plus plus

Sat, Apr 30 2011 JoungKyun.Kim <http://oops.org>
	- enhanced krisp.m4

Fri, Apr 22 2011 JoungKyun.Kim <http://oops.org>
	- changed build environment with automake
	- revert gcc visibility


* 3.1.2
Fri, Jan 14 2011 JoungKyun.Kim <http://oops.org>
	- release 3.1.2
	- fixed segfault cause of miss casting

Tue, Jan 12 2011 JoungKyun.Kim <http://oops.org>
	- add english man pages
	- debianized

Mon, Nov  8 2010 JoungKyun.Kim <http://oops.org>
	- fixed build environments on other OS (ulong type, make command)
	- 3.1.2 dev branch start


* 3.1.1
Mon, Sep 27 2010 JoungKyun.Kim <http://oops.org>
	- release 3.1.1

Fri, Sep 24 2010 JoungKyun.Kim <http://oops.org>
	- add journal_mode=OFF option on sqlite3 database handle
	- add _REENTRANT compile flag
	- changed default value of db_time_stamp_interval from 3600 sec to 0 sec

Mon, Sep 20 2010 JoungKyun.Kim <http://oops.org>
	- initialize db handler to NULL when database connect failed on kr_open


* 3.1.0
Fri, Sep 10 2010 JoungKyun.Kim <http://oops.org>
	- changed kr_close api proto type
	  kr_close (KR_API *) => kr_close (KR_API **)
	- reconnect database when changed mtime of database
	  add 4 members in KR_NET struct
	  database
	  db_time_stamp_interval
	  db_time_stamp
	  db_stamp_checked


* 3.0.3

Fri, Sep 10 2010 JoungKyun.Kim <http://oops.org>
	- fixed segfault kr_search(_ex) when KR_NET structure is null

Thu, Aug 12 2010 JoungKyun.Kim <http://oops.org>
	- fixed typo


* 3.0.2

Sun, Aug  8 2010 JoungKyun.Kim <http://oops.org>
	- release 3.0.2
	- link with libipcalc
	- remove ip2long, netcalc that forked libipcalc
	- add --with-compat build option for makeing compat library
	- add krisp.pc that is pkg-config configuration file


* 3.0.1

Tue, Aug  3 2010 JoungKyun.Kim <http://oops.org>
	- release 3.0.1
	- moved header files to %includedir/krisp
	- removed dependency library on --libs and, add --link for dependency
	  library on krisp-config


* 3.0.0

Fri, Jul 16 2010 JoungKyun.Kim <http://oops.org>
	- release 3.0.0
	- update database on july

Fri, Jul  2 2010 JoungKyun.Kim <http://oops.org>
	- enhanced build environments

Wed, Jun 30 2010 JoungKyun.Kim <http://oops.org>
	- add kr_long2ip_r api for thread safe
	- changed mutex lock per db constructor

Tue, Jun 29 2010 JoungKyun.Kim <http://oops.org>
	- rewrite _kr_valid_ip_address and change proto type

Sun, Jun 27 2010 JoungKyun.Kim <http://oops.org>
	- initialize verbose member of KRNET_API(_ex) on kr_search(_ex) API

Fri, Jun 25 2010 JoungKyun.Kim <http://oops.org>
	- rename internal api name
	- make local symbol
	- fix wrong caculation of netcalc when use FQDN instead of IP address

Thu, Jun 24 2010 JoungKyun.Kim <http://oops.org>
	- support multiple build (make -j4)
	- fixed dl link error

Fri, Jun 18 2010 JoungKyun.Kim <http://oops.org>
	- fixed sqlite3_open_v2 compatibility problems under sqlite 3.5.0
	- add bool type to change some APIs and variables and structure members
	- don't call kr_close api any more when failed kr_open

Wed, Jun 16 2010 JoungKyun.Kim <http://oops.org>
	- changed proto type of kr_open
	- add err member on kr_search/kr_search_ex for thread safe

Mon, Jun 14 2010 JoungKyun.Kim <http://oops.org>
	- add thread safe

Thu, Jun 10 2010 JoungKyun.Kim <http://oops.org>
	- add database file check on kr_open

Wed, Jun  9 2010 JoungKyun.Kim <http://oops.org>
	- fixed segfault on kr_open

Tue, Jun  8 2010 JoungKyun.Kim <http://oops.org>
	- init verbose member of KR_API to 0 on kr_open
	- add dummy member array size on RAW_KRNET_API
	- changed frontend api from function to define

Sat, May 22 2010 JoungKyun.Kim <http://oops.org>
	- changed query algoritm
	- changed database structure
	- support internal geoip/geoipisp database
	  Don't need geoip library any more
	  Support convert scripts
	- update SONAME to 2.0.0
	  changed usage of almost APIs




netcalc.c
/*
 * $Id: netcalc.c,v 1.7 2010-06-24 16:52:38 oops Exp $
 */

#include <krisp.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#define PNAME "netcalc"
#define CIDS 0
#define IPv4 1

#ifdef HAVE_GETOPT_LONG
static struct option long_options [] = { // {{{
	/* Options without arguments: */
	{ "broadcast",    no_argument,       NULL, 'b' },
	{ "help",         no_argument,       NULL, 'h' },
	{ "mask",         no_argument,       NULL, 'm' },
	{ "network",      no_argument,       NULL, 'n' },
	{ "prefix",       no_argument,       NULL, 'p' },
	{ "shell-format", no_argument,       NULL, 's' },
	{ "verbose",      no_argument,       NULL, 'v' },

	/* Options accepting an argument: */
	//{ "datafile",   required_argument, NULL, 'f' },
	{ 0, 0, 0, 0 }
}; // }}}
#endif

void usage (char * prog) { // {{{
	fprintf (stderr, "%s v%s: get network informations\n", prog, KRISP_VERSION);
	fprintf (stderr, "Usage: %s [option] ipaddress/[mask|prefix]\n", prog);
	fprintf (stderr, "       %s [option] StartIP EndIP\n", prog);
	fprintf (stderr, "Options:\n");
	fprintf (stderr, "         -h, --help             print this message\n");
	fprintf (stderr, "         -b, --broadcast        print broadcast\n");
	fprintf (stderr, "         -n, --network          print network\n");
	fprintf (stderr, "         -m, --mask             print network mask\n");
	fprintf (stderr, "         -p, --prefix           print network prefix\n");
	fprintf (stderr, "         -s, --shell-format     print shell format\n");
	fprintf (stderr, "         -a, --all              print all\n");

	exit (1);
} // }}}

typedef struct opt_value {
	short	broadcast;
	short	mask;
	short	network;
	short	prefix;
	short	shell;
} OPTV;

typedef struct resutls_value {
	ulong	start;
	ulong	end;
	short	prefix;
	ulong	mask;
	ulong	network;
	ulong	broadcast;
} RETV;

int main (int argc, char ** argv) {
	char 	ip[256] = { 0, };
	char *	postfix;
	char *	pformat;
	int		opt;
	short	optno;
	bool	verbose = false;
	OPTV	optv;
	RETV	r;

	optv.broadcast = 0;
	optv.mask      = 0;
	optv.network   = 0;
	optv.prefix    = 0;
	optv.shell     = 0;

#ifdef HAVE_GETOPT_LONG
	while ( (opt = getopt_long (argc, argv, "abhnmpsv", long_options, (int *) 0)) != EOF ) {
#else
	while ( (opt = getopt (argc, argv, "abhnmpsv")) != EOF ) {
#endif
		switch (opt) {
			case 'v' :
				set_true (verbose);
				break;
			case 'a' :
				optv.broadcast++;
				optv.mask++;
				optv.network++;
				optv.prefix++;
				break;
			case 'b' :
				optv.broadcast++;
				break;
			case 'm' :
				optv.mask++;
				break;
			case 'n' :
				optv.network++;
				break;
			case 'p' :
				optv.prefix++;
				break;
			case 's' :
				optv.shell++;
				break;
			default:
				usage (PNAME);
		}
	}

	if ( optv.shell )
		verbose = false;

	optno = argc - optind;
	if ( optno < 1 || optno > 2 ) {
		usage (PNAME);
		return 1;
	}

	if ( optno == 1 ) {
		/*
		 * Case IP/Prefix or IP/Netmask
		 */
		char *	mask;

		SAFECPY_256 (ip, argv[optind]);
		if ( (mask = strchr (ip, '/')) == NULL ) {
			fprintf (stderr, "ERROR: %s format is IPADDRESS/[NETMASK|PREFIX]\n", argv[1]);
			usage (PNAME);
			return 1;
		}
		*mask = 0;

		if ( valid_ipv4_addr (ip) ) {
			fprintf (stderr, "ERROR: %s is not IPv4 nor Long IP address\n", ip);
			return 1;
		}
		r.start = kr_ip2long (ip);

		if ( strchr (mask + 1, '.') ) {
			char			_mask[256];

			SAFECPY_256 (_mask, mask + 1);
			if ( valid_ipv4_addr (_mask) ) {
				fprintf (stderr, "ERROR: %s is not IPv4 nor Long IP address\n", _mask);
				return 1;
			}

			r.mask = kr_ip2long (_mask);
			r.prefix = kr_long2prefix (r.mask);
		} else {
			r.prefix = atoi (mask + 1);
			if ( r.prefix > 32 || r.prefix < 0 ) {
				fprintf (stderr,
						"ERROR: %s is overflow or underflow on prefix range.\n",
						mask + 1
				);
				return 1;
			}
			r.mask = kr_prefix2long (r.prefix);
		}

		r.network = kr_network (r.start, r.mask);
		r.broadcast = r.end = kr_broadcast (r.start, r.mask);
	} else {
		/*
		 * Case given range from start ip to end ip
		 */
		if ( strchr (argv[optind], '/') != NULL ||
			strchr (argv[optind + 1], '/') != NULL ) {
			fprintf (
				stderr,
				"ERROR: %s and %s is must ip address\n",
				argv[optind],
				argv[optind + 1]
			);
			usage (PNAME);
			return 1;
		}

		SAFECPY_256 (ip, argv[optind]);
		if ( valid_ipv4_addr (ip) ) {
			fprintf (stderr, "ERROR: %s is not IPv4 nor Long IP address\n", ip);
			return 1;
		}
		r.start = kr_ip2long (ip);

		SAFECPY_256 (ip, argv[optind + 1]);
		if ( valid_ipv4_addr (ip) ) {
			fprintf (stderr, "ERROR: %s is not IPv4 nor Long IP address\n", ip);
			return 1;
		}
		r.end = kr_ip2long (ip);

		if ( r.end < r.start ) {
			r.mask = r.start;
			r.start = r.end;
			r.end = r.mask;
		}

		r.prefix = kr_prefix (r.start, r.end);
		r.mask = kr_prefix2long (r.prefix);
		r.network = kr_network (r.start, r.mask);
		r.broadcast= kr_broadcast (r.start, r.mask);
	}

	postfix = optv.shell ? "=" : " : ";
	pformat = optv.shell ? "%s%s" : "%-11s%s";

	if ( verbose ) {
		printf ("%-11s : %s - ", "GIVEN RANGE", kr_long2ip (r.start));
		printf ("%s\n", kr_long2ip (r.end));
	}

	if ( optv.network ) {
		if ( verbose || optv.shell )
			printf (pformat, "NETWORK", postfix);

		printf ("%s\n", kr_long2ip (r.network));
	}

	if ( optv.broadcast ) {
		if ( verbose || optv.shell )
			printf (pformat, "BROADCAST", postfix);

		printf ("%s\n", kr_long2ip (r.broadcast));
	}

	if ( optv.mask ) {
		if ( verbose || optv.shell )
			printf (pformat, "NETMASK", postfix);

		printf ("%s\n", kr_long2ip (r.mask));
	}

	if ( optv.prefix ) {
		if ( verbose || optv.shell )
			printf (pformat, "PREFIX", postfix);

		printf ("%d\n", r.prefix);
	}

	return 0;
}

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */
